pkg/controller/node/nodecontroller.go

```diff
@@ -910,3 +1033,77 @@ func (nc *NodeController) ComputeZoneState(nodeReadyConditions []*api.NodeCondit
 		return notReadyNodes, stateNormal
 	}
 }
+
+func tryAddTaintToNode(kubeClient clientset.Interface, nodeName string, taintToAdd api.Taint) (bool, error) {
+	node, err := kubeClient.Core().Nodes().Get(nodeName)
+	if err != nil {
+		return false, err
+	}
+
+	oldTaints, err := api.GetTaintsFromNodeAnnotations(node.Annotations)
+	if err != nil {
+		return false, err
+	}
+
+	for _, oldTaint := range oldTaints {
+		if taintToAdd.MatchTaint(oldTaint) {
+			return false, nil
+		}
+	}
+
+	newTaints := append(oldTaints, taintToAdd)
+	taintsData, err := json.Marshal(newTaints)
+	if err != nil {
+		return false, err
+	}
+
+	if len(node.Annotations) == 0 {
+		node.Annotations = map[string]string{}
+	}
+	node.Annotations[api.TaintsAnnotationKey] = string(taintsData)
+
+	_, err = kubeClient.Core().Nodes().Update(node)
+	return err == nil, err
+}
+
+// tryRemoveTaintsOffNode tries to remove taints off a node,
+// return true if taints removed, or return false if taints don't exist.
+func tryRemoveTaintsOffNode(kubeClient clientset.Interface, nodeName string, taintsToRemove ...api.Taint) (bool, error) {
```

Comment:
I would prefer that we delay refactoring for code compactness. Let's get this PR into a state where it actually works and implements all the functionality, and then clean it up.